### ISSUE TITLE
Improved performance on Parser

### ISSUE THREAD

**Issue by RupertAvery** [Maximilian-Winter/SrslBytecodeVmAndCodeGenerator#9]
So, I was right about the exception throwing slowing down the parser.

I managed to rewrite a whole lot of the parsing to return a Context value along with the node, and a flag if there was a failure. This flag gets handled in places where the exception should have been handled.

**Original Code - Release Mode**

```
Parsing completed in 12610ms
14930352
--Elapsed Time for Interpreting Run 0 is 309 ms
```

**New Code - Release Mode**

```
Parsing completed in 54ms
14930352
--Elapsed Time for Interpreting Run 0 is 358 ms
```

See https://github.com/Maximilian-Winter/SrslBytecodeVmAndCodeGenerator/pull/8

I also separated the rule functions e.g. rule() from the implementation _rule() in separate partial classes to clean up.

**Comment by Maximilian-Winter** [comment]
Thank you very much for your commitmend and your general Input, you are great!

**Comment by RupertAvery** [comment]
Thanks for merging!

Merged with Pull request #8
